pkg/apis/core/validation/validation.go

```diff
@@ -6735,9 +6735,20 @@ func validateResourceClaimNames(claims []core.ResourceClaim, podClaimNames sets.
 			allErrs = append(allErrs, field.Required(fldPath.Index(i), ""))
 		} else {
 			if names.Has(name) {
+				// All requests of that claim already referenced.
 				allErrs = append(allErrs, field.Duplicate(fldPath.Index(i), name))
 			} else {
-				names.Insert(name)
+				key := name
```

Comment:
You are reading this right. The rationale is that all requests of `claimA` are already made available. The second list entry is redundant.

That's the same rationale why previously, this redundancy wasn't allowed:
```
- name: claimA
- name: claimA
```

Reply:
> Shouldn't we error in both cases?

We do? But perhaps I misunderstood.

This fails in validation (pre-1.31):
```
- name: claimA
- name: claimA
```

This now also fails (1.31):
```
- name: claimA
- name: claimA
  request: req-0
```

This passes:
```
- name: claimA
  request: req-0
- name: claimA
  request: req-1
```

> and both modes should be in tests

They are:
```
	successCases := map[string]*core.Pod{
	...
		"multiple claims": podtest.MakePod("",
			podtest.SetContainers(podtest.MakeContainer("ctr", podtest.SetContainerResources(core.ResourceRequirements{Claims: []core.ResourceClaim{{Name: "my-claim"}, {Name: "another-claim"}}}))),
...
		"multiple claims with requests": podtest.MakePod("",
			podtest.SetContainers(podtest.MakeContainer("ctr", podtest.SetContainerResources(core.ResourceRequirements{Claims: []core.ResourceClaim{{Name: "my-claim", Request: requestName}, {Name: "another-claim", Request: requestName}}}))),
...
	failureCases := map[string]*core.Pod{
...
		"pod claim name duplicates": podtest.MakePod("",
			podtest.SetContainers(podtest.MakeContainer("ctr", podtest.SetContainerResources(core.ResourceRequirements{Claims: []core.ResourceClaim{{Name: "my-claim"}, {Name: "my-claim"}}}))),
...
		"pod claim name duplicates without and with request": podtest.MakePod("",
			podtest.SetContainers(podtest.MakeContainer("ctr", podtest.SetContainerResources(core.ResourceRequirements{Claims: []core.ResourceClaim{{Name: "my-claim"}, {Name: "my-claim", Request: "req-0"}}}))),
...
		"pod claim name duplicates with requests": podtest.MakePod("",
			podtest.SetContainers(podtest.MakeContainer("ctr", podtest.SetContainerResources(core.ResourceRequirements{Claims: []core.ResourceClaim{{Name: "my-claim", Request: "req-0"}, {Name: "my-claim", Request: "req-0"}}}))),
```

This is with a small typo fix ("multiple claim" -> "multiple claims").

One failure test case was missing:
```
		"pod claim name duplicates with and without request": podtest.MakePod("",
			podtest.SetContainers(podtest.MakeContainer("ctr", podtest.SetContainerResources(core.ResourceRequirements{Claims: []core.ResourceClaim{{Name: "my-claim", Request: "req-0"}, {Name: "my-claim"}}}))),
```

That wasn't handled as intended because the key in the set is "my-claim/req-0" when checking the second "my-claim".  Fixed.

What cannot be validated is whether the request names are really available in the referenced claim. That depends on knowing what the claim will be.

Reply:
Does this fail?

```yaml
- name: claimA
  request: requestA
- name: claimA
```

Tim and I think it should...

Reply:
Agree with John - it seems likely to be a mistake, better to demand the user clean it up.

Reply:
But that's what I am saying?! I agree that this should fail and it does with the current validation:

```
- name: claimA
  request: requestA
- name: claimA
```

Reply:
Ah! Yes, you fixed it. I see you note that above. That was exactly the case I was calling out, and now you have added the test and fixed it. So you can resolve this.